torch/optim/_functional.py

```diff
@@ -357,3 +357,48 @@ def asgd(params: List[Tensor],
             ax.add_(param.sub(ax).mul(mu))
         else:
             ax.copy_(param)
+
+
+def nadam(params: List[Tensor],
+          grads: List[Tensor],
+          exp_avgs: List[Tensor],
+          exp_avg_sqs: List[Tensor],
+          m_schedules: List[float],
+          state_steps: List[int],
+          *,
+          beta1: float,
+          beta2: float,
+          lr: float,
+          weight_decay: float,
+          schedule_decay: float,
+          eps: float):
+    r"""Functional API that performs NAdam algorithm computation.
+
+    See :class:`~torch.optim.NAdam` for details.
+    """
+
+    for i, param in enumerate(params):
+        grad = grads[i]
+        exp_avg = exp_avgs[i]
+        exp_avg_sq = exp_avg_sqs[i]
+        m_schedule = m_schedules[i]
+        step = state_steps[i]
+
+        bias_correction2 = 1 - beta2 ** step
+
+        if weight_decay != 0:
+            grad = grad.add(param, alpha=weight_decay)
+
+        # calculate the momentum cache \mu^{t} and \mu^{t+1}
+        momentum_cache_t = beta1 * (1. - 0.5 * (0.96 ** (step * schedule_decay)))
+        momentum_cache_t_1 = beta1 * (1. - 0.5 * (0.96 ** ((step + 1) * schedule_decay)))
+        m_schedule_new = m_schedule * momentum_cache_t
+        m_schedule_next = m_schedule * momentum_cache_t * momentum_cache_t_1
+
+        # decay the first and second moment running average coefficient
+        exp_avg.mul_(beta1).add_(grad, alpha=1 - beta1)
+        exp_avg_sq.mul_(beta2).addcmul_(grad, grad, value=1 - beta2)
+
+        denom = exp_avg_sq.div(bias_correction2).sqrt().add_(eps)
+        param.addcdiv_(grad, denom, value=-lr * (1. - momentum_cache_t) / (1. - m_schedule_new))
+        param.addcdiv_(exp_avg, denom, value=-lr * momentum_cache_t_1 / (1. - m_schedule_next))
```

Comment:
The notation in the [paper](https://arxiv.org/pdf/1412.6980.pdf) is significantly different. Do you follow a difference implementation? If you've derived this yourself, can you provide the derivation?

I see that you've tried to stay close to the adam implementation here in pytorch. :)
<img width="1704" alt="Screen Shot 2021-06-15 at 11 44 25 AM" src="https://user-images.githubusercontent.com/3047868/122084047-70d71400-cdcf-11eb-9dcc-11b8d6d606bd.png">

Reply:
Yes :)  I actually thought it would be nice if all the algorithms follows similar notations/patterns in our codebase, so it would be much more easy to read them. I would be glad to derive in a videochat or i can make a write-up/quip or pdf ready for derivation.

Reply:
as discussed offline, please change the name of the variables to make them consistent with the paper and themselves:
```
momentum_cache_t -> mu
momentum_cache_t_1 -> mu_next
m_schedule_new -> mu_product
m_schedule_next -> mu_product_next
```

Reply:
its done ;)